bindata/network/openshift-sdn/alert-rules.yaml

```diff
@@ -39,40 +42,39 @@ spec:
-        histogram_quantile(.95, kubeproxy_sync_proxy_rules_duration_seconds_bucket) * on(pod) group_right kube_pod_info{namespace="openshift-sdn",  pod=~"sdn-[^-]*"} > 15
+        histogram_quantile(.95, kubeproxy_sync_proxy_rules_duration_seconds_bucket) 
+        * on(namespace, pod) group_right topk by (namespace, pod) (1, kube_pod_info{namespace="openshift-sdn",  pod=~"sdn-[^-]*"}) > 15
       labels:
         severity: warning
-    - alert: IPTableSyncCluster
+    - alert: ClusterProxyApplySlow
       annotations:
-        message: The average time for SDN pods to sync iptables is too high.
+        message: The cluster is taking too long, on average, to apply kubernetes service rules to iptables.
       expr: |
         histogram_quantile(0.95, sum(rate(kubeproxy_sync_proxy_rules_duration_seconds_bucket[5m])) by (le)) > 10
       labels:
         severity: warning
-    - alert: NodeIPTablesStale
-    # there is some scrape delay and some other offset 120 is not really 120s but it is still too long
+    - alert: NodeProxyApplyStale
       annotations:
-        message: SDN pod {{"{{"}} $labels.pod {{"}}"}} on node {{"{{"}} $labels.node {{"}}"}} has gone too long without syncing iptables rules.
+        message: SDN pod {{"{{"}} $labels.pod {{"}}"}} on node {{"{{"}} $labels.node {{"}}"}} has stale kubernetes service rules in iptables.
+      # Query: find pods where
+      #  - the queued-timestamp is at least 30 seconds after the applied-timestamp
+      #  - the pod (still) exists
       expr: |
-            (timestamp(kubeproxy_sync_proxy_rules_last_timestamp_seconds)
-            - on(pod) kubeproxy_sync_proxy_rules_last_timestamp_seconds)
-            * on(pod) group_right kube_pod_info{namespace="openshift-sdn",  pod=~"sdn-[^-]*"}> 120
-      for: 20m
+        (kubeproxy_sync_proxy_rules_last_queued_timestamp_seconds - kubeproxy_sync_proxy_rules_last_timestamp_seconds)
+        * on(namespace, pod) group_right() topk by (namespace, pod) (1, kube_pod_info{namespace="openshift-sdn",pod=~"sdn-[^-]*"})
+        > 30
+      for: 5m # quiet any churn on sdn restarts.
       labels:
         severity: warning
-    - alert: ClusterIPTablesStale
-    # there is some scrape delay and some other offset 90 i not really 90s but it is still too long
+    - alert: SDNPodNotReady
       annotations:
-        message: The average time between iptables resyncs is too high. NOTE - There is some scrape delay and other offsets, 90s isn't exact but it is still too high.
+        message: SDN pod {{"{{"}} $labels.pod {{"}}"}} on node {{"{{"}} $labels.node {{"}}"}} is not ready.
       expr: |
-        quantile(0.95,
-            timestamp(kubeproxy_sync_proxy_rules_last_timestamp_seconds)
-            - on(pod) kubeproxy_sync_proxy_rules_last_timestamp_seconds
-            * on(pod) group_right kube_pod_info{namespace="openshift-sdn",  pod=~"sdn-[^-]*"}) > 90
-      for: 20m
+        kube_pod_status_ready{namespace='openshift-sdn', condition='true'} == 0
```

Comment:
This will fail to trigger on a master if sdn-controller is running but sdn is not.

Reply:
:facepalm: crap.

Reply:
No, this doesn't join on nodes; any non-ready pod will get an alert. It should be correct.